### PR TITLE
[libbeat] Add config options for mutex/block/mem profiling rates

### LINKED ISSUE #33572: [libbeat] Add support for block and mutex profiling
**Describe the enhancement:**

Add a configuration option to Beats to enable block profiling and mutex profiling.

**Describe a specific use case for the enhancement or feature:**

This would be used to help identify performance issues. Block profile shows where goroutines block waiting on synchronization primitives (including timer channels). Mutex profile reports the lock contentions. When you think your CPU is not fully utilized due to a mutex contention, you can use this.

### References

- https://github.com/DataDog/go-profiler-notes/blob/main/block.md
- https://rakyll.org/mutexprofile/
- https://go.dev/doc/diagnostics
- [runtime.SetBlockProfileRate](https://golang.org/pkg/runtime/#SetBlockProfileRate)
- [runtime.SetMutexProfileFraction](https://golang.org/pkg/runtime/#SetMutexProfileFraction)
- https://github.com/elastic/apm-server/commit/9f185fe0515c479c5eac46f17c0d04d414ddeaea (apm-server already has this)

### PR DESCRIPTION
**Pull request by andrewkroh**
## What does this PR do?

This enables performing mutex and block profiling. It also allows for configuring the mem profiling rate. The new settings are:

- `http.pprof.block_profile_rate`
- `http.pprof.mem_profile_rate`
- `http.pprof.mutex_profile_rate`

Closes #33572

## Why is it important?

It makes it possible to do mutex and block profiling.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #33572

### CODE REVIEW

File: libbeat/docs/http-endpoint.asciidoc [hunk 1]
```diff
@@ -33,6 +33,21 @@ current user.
...
+`http.pprof.mem_profile_rate`:: (Optional) `mem_profile_rate` controls the
+fraction of memory allocations that are recorded and reported in the memory
+profile. The profiler aims to sample an average of one allocation per
+`mem_profile_rate` bytes allocated. To include every allocated block in the
+profile, set `mem_profile_rate` to 1. To turn off profiling entirely, set
+`mem_profile_rate` to 0.
+`http.pprof.mutex_profile_rate`:: (Optional) `mutex_profile_rate` controls the
+fraction of mutex contention events that are reported in the mutex profile. On
+average 1/rate events are reported. The previous rate is returned. To turn off
+profiling entirely, pass rate 0. To just read the current rate, pass rate < 0.
```
Comment (efd6) [inline libbeat/docs/http-endpoint.asciidoc:50]: Is it worth noting the endpoint for each of these?

File: libbeat/pprof/pprof.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+// IsEnabled returns true if the pprof config is non-nil and either 'enabled'
+// is not set or it is set to true.
+func (c *Config) IsEnabled() bool {
+	return c != nil && (c.Enabled == nil || *c.Enabled)
+}
+
+// SetRuntimeProfilingParameters pushes the configuration parameters into the Go runtime.
+// Profiling rates should be set once, early on in the program.
+func SetRuntimeProfilingParameters(cfg *Config) {
```
Comment (cmacknz) [inline libbeat/pprof/pprof.go:48]: I don't see this used anywhere, unless I'm missing something.

**Review by cmacknz — APPROVED** [review]
LGTM me. 

FYI when the agent v2 branch is merged, the v2 protocol client linked into the Beats will support the entire set of pprof diagnostics by default _without_ having to enable `http.pprof` first. The profiles aren't fetched over the pprof HTTP interface anymore.

This is the default set of profiles collected from `elastic-agent diagnostics collect --pprof` https://github.com/elastic/elastic-agent-client/blob/bc376a4e0f9f7677b30770c69dd6c80437924e91/pkg/client/client_v2.go#L658-L665

File: libbeat/pprof/pprof.go (hunk 2 above)
Comment (andrewkroh) [reply libbeat/pprof/pprof.go:48]: Absolutely should be. I guess I'm not as good at multitasking as I thought. 😢

**Comment by andrewkroh** [comment]
> linked into the Beats will support the entire set of pprof diagnostics by default without having to enable http.pprof first

Nice. But I don't see anywhere that the elastic-agent-client is calling `runtime.SetBlockProfileRate()` or the `runtime.SetMutexProfileFraction()`. So even though it can collect those profiles they will be empty.

**Comment by cmacknz** [comment]
> But I don't see anywhere that the elastic-agent-client is calling runtime.SetBlockProfileRate() or the runtime.SetMutexProfileFraction(). So even though it can collect those profiles they will be empty.

Thanks I'll make a note to set these up when running under agent.

File: libbeat/docs/http-endpoint.asciidoc (hunk 1 above)
Comment (andrewkroh) [reply libbeat/docs/http-endpoint.asciidoc]: Sounds like a good idea. Added.

File: libbeat/docs/http-endpoint.asciidoc [hunk 3]
```diff
@@ -33,6 +33,23 @@ current user.
...
+`http.pprof.mem_profile_rate`:: (Optional) `mem_profile_rate` controls the
+fraction of memory allocations that are recorded and reported in the memory
+profile available from `/debug/pprof/heap`. The profiler aims to sample an
+average of one allocation per `mem_profile_rate` bytes allocated. To include
+every allocated block in the profile, set `mem_profile_rate` to 1. To turn off
+profiling entirely, set `mem_profile_rate` to 0.
+`http.pprof.mutex_profile_rate`:: (Optional) `mutex_profile_rate` controls the
+fraction of mutex contention events that are reported in the mutex profile
+available from `/debug/pprof/mutex`. On average 1/rate events are reported. The
+previous rate is returned. To turn off profiling entirely, pass rate 0. To just
```
Comment (andrewkroh) [inline libbeat/docs/http-endpoint.asciidoc:51]: I need to remove that last sentence about reading from each. And mention the defaults for each.

Plus these three needed added to the reference config file.

**Review by efd6 — APPROVED** [review]

**Pull request merged**